godot-core/src/builtin/array.rs

```diff
@@ -266,8 +306,9 @@ impl<T: GodotType> Array<T> {
     /// To create a deep copy, use [`duplicate_deep()`][Self::duplicate_deep] instead.
     /// To create a new reference to the same array data, use [`clone()`][Clone::clone].
     pub fn duplicate_shallow(&self) -> Self {
-        let duplicate: VariantArray = self.as_inner().duplicate(false);
-        // SAFETY: duplicate() returns a typed array with the same type as Self
+        // SAFETY: We never write to the duplicated array, and all values read are read as `Variant`.
+        let duplicate: VariantArray = unsafe { self.as_inner_ref().duplicate(false) };
+        // SAFETY: duplicate() returns a typed array with the same type as Self, and all values are taken from `self` so have the right type.
```

Comment:
An empty line before `// SAFETY` would help group the lines.

(lots of similar cases)